core/io/pread_spec.rb

```diff
@@ -26,11 +26,73 @@
       buffer.should == "567"
     end
 
+    it "shrinks the buffer in case of less bytes read" do
+      buffer = "foo"
+      @file.pread(1, 0, buffer)
+      buffer.should == "1"
+    end
+
+    it "grows the buffer in case of more bytes read" do
+      buffer = "foo"
+      @file.pread(5, 0, buffer)
+      buffer.should == "12345"
+    end
+
     it "does not advance the file pointer" do
       @file.pread(4, 0).should == "1234"
       @file.read.should == "1234567890"
     end
 
+    it "returns an empty string for maxlen = 0, regardless of the offset" do
+      @file.pread(0, 4).should == ""
+      @file.pread(0, 400).should == ""
+    end
+
+    it "does not reset the buffer when reading with maxlen = 0" do
+      buffer = "foo"
+      @file.pread(0, 4, buffer)
+      buffer.should == "foo"
+
+      @file.pread(0, 400, buffer)
+      buffer.should == "foo"
+    end
+
+    it "ignores the current offset" do
+      @file.pos = 3
+      @file.pread(4, 0).should == "1234"
+    end
+
+    it "converts maxlen and offset to Integers using #to_int" do
+      maxlen = mock('maxlen')
+      maxlen.should_receive(:to_int).and_return(4)
+      offset = mock('offset')
+      offset.should_receive(:to_int).and_return(0)
+      @file.pread(maxlen, offset).should == "1234"
+    end
+
+    it "converts a buffer to String using to_str (but you cannot use this value)" do
+      buffer = mock('buffer')
+      buffer.should_receive(:to_str).at_least(1).and_return("foo")
+      @file.pread(0, 4, buffer).should == "foo"
+      buffer.should_not.is_a?(String)
+      buffer.to_str.should == "foo"
+    end
+
+    it "expects Integer values for maxlen and offset" do
+      -> { @file.pread("maxlen", 0) }.should raise_error(TypeError)
+      -> { @file.pread(4, "offset") }.should raise_error(TypeError)
+    end
```

Comment:
**suggestion:** it's an edge case when object doesn't respond to `#to_int` method, so I would reflect it in the test title (as well as in the test for non-String buffer).

Reply:
It turns out the error message in this case is different between linux and mac os (https://github.com/ruby/spec/actions/runs/6315519202/job/17148232880?pr=1079). What would be the best course of action here, matching `no implicit conversion` as the maximal shared string, or matching it with a Regex union of both messages?

And honestly, I think it's kind of weird that a non-int type for maxlen results in a different error message than offset, I doubt this is the intended behaviour.

Reply:
I would suggest to use something other than `String` literal, e.g. `Object.new`. This way error message should be the same in both cases.

It's caused by using different C-functions to implicitly convert Ruby object to some type. One of them (`rb_num2ll` - to convert to `long long`) has a case for `String` object that leads to a specific error message. Other similar functions (to convert to `long` and `int`) don't have such case. Choosing what function to use depends on C numeric type sizes that is related to a compiler and C standard library implementation (and OS of course):

```c
#ifdef OFFT2NUM
# /* take that. */
#elif SIZEOF_OFF_T == SIZEOF_LONG_LONG
# define OFFT2NUM RB_LL2NUM
#elif SIZEOF_OFF_T == SIZEOF_LONG
# define OFFT2NUM RB_LONG2NUM
#else
# define OFFT2NUM RB_INT2NUM
#endif
```